### ISSUE TITLE
Convert old GPU ops to GpuArray ops

### ISSUE THREAD

**Issue by nouiz** [Theano/Theano#1471]
This is a fast conversion. We do not support OpenCL to don't maintain 2 code base for a long time. OpenCL is for later. We need help from the community for this.

How to convert.
0) see http://www.deeplearning.net/software/theano/tutorial/extending_theano.html to get an idea of how to make an op.
1) Install compyte (a new dependency for Theano). Use the version in abergeron account in the reorg branch: http://deeplearning.net/software/compyte/installation.html
1) Fast way(but no OpenCL support), copy the old Op from the CudaNdarray back-end under theano/sandbox/cuda/*.py and its tests to theano/sandbox/gpuarray[/tests]. See gh-1590 and gh-1634 for examples.
   Then modify it, 
      - by adding the c_compilers, c_headers and if needed c_support_code, c_init_code methods.
      - modify the make node to return an GpuArrayVariabe with GpuArrayType and make sure to call as_gpu_array_varible(input) on all input variable.
      - Modify the CudaNdarray macro the the compatible version: https://github.com/abergeron/compyte/blob/reorg/src/compyte/numpy_compat.h
      - The macro CudaNdarray_DEV_DATA is replaced by the fct call cuda_get_ptr() defined in c_support_code, the offset need to be added: (float *)(((char *)cuda_get_ptr(%(inp)s->ga.data))+%(inp)s->ga.offset)
      - Don't forget that when changing the macro CudaNdarray_HOST_STRIDES to PyGpuArray_STRIDES(), the unit change from number of element to number of bytes. So you need to divide it by 4 if you support only float32. Otherwise, you need to update the kernel code.
      - change NUM_VECTOR_OP_THREADS_PER_BLOCK to 256
      - change CNDA_THREAD_SYNC to """        if config.gpuarray.sync:
            code += "GpuArray_sync(&%(zz)s->ga);\n" % dict(zz=zz)"""
      - Make sure all tests pass. Commit.
      - Make a new optimizer function by reusing the new op_lifter.
   Add support for more dtype then float32 and modify tests for this.
1.1) Optional. Make the new op inherit from the CPU op to share code.
1.2) Optional. Make the CPU tests generic and merge it in the gpu tests. Reuse for the GPU tests.

note::
- It can happen that after nosetests results are printed that it have a segfault or another errors. This is due to the order where we close stuff. Just ignore it for now, we will take care of it. This is not related to Ops code.

Absolutely needed ops:
- [x] HostFromGpu, GpuFromHost, GpuFromCuda, CudaFromGpu

This is the list of ops taken from the DLT that need to be converted (can give an idea of the order to convert):
- logistic_sgd

| Op | CUDA | OpenCL | float16 | Note |
| --- | --- | --- | --- | --- |
| Dot22 | Yes | Yes | yes | float16 need cuda 7.5 |
| Gemm | Yes | Yes | yes | float16 need cuda 7.5 |
| MaxAndArgmax |  |  |  | old back-end do not have it |
| Elemwise | c code | python, no dim collapsins | Yes |  |
| Subtensor | yes | yes | yes |  |
| DimShuffle | yes | yes | yes |  |
| shape_i | yes | yes | yes |  |
| Sum(Careduce) | c code from old  back-end | python new code | yes (with no pre_scalar_op) |  |
| CrossentropySoftmaxArgmax1HotWithBias | yes | yes | yes |  |
| CrossentropySoftmax1HotWithBiasDx | yes | yes | yes |  |
| Alloc/AllocEmpty | yes | yes | yes |  |
- logistic_cg

| Op | CUDA | OpenCL | float16 | Note |
| --- | --- | --- | --- | --- |
| Reshape | Yes | Yes | yes |  |
| SoftmaxWithBias | Yes | yes | yes |  |
| IncSubtensor | yes | yes | yes |  |
- mlp
  - ops above
- conv mlp

| Op | CUDA | OpenCL | float16 | Note |
| --- | --- | --- | --- | --- |
| ConvOp | Yes | no | no |  |
| DownsampleFactorMax |  |  |  | cudnn give it |
| Downsamplefactormaxgrad |  |  |  | cudnn give it |
| Join | Yes | Yes | Yes |  |
| Rebroadcast | Yes | Yes | Yes |  |
| Prod |  |  |  | yes with sum, see sum for info |
- test da and sda

| Op | CUDA | OpenCL | float16 | Note |
| --- | --- | --- | --- | --- |
| MRG | Yes | yes | yes |  |
- dbn

| Op | CUDA | OpenCL | float16 | Note |
| --- | --- | --- | --- | --- |
| GEMV | Yes | yes | no |  |
- rbm 

| Op | CUDA | OpenCL | float16 | Note |
| --- | --- | --- | --- | --- |
| GEMV | Yes | yes | no |  |
| DeeCopy | Yes | Yes | Yes |  |
- rnnrbm

Other Theano gpu op needed:

| Op | CUDA | OpenCL | float16 | Note |
| --- | --- | --- | --- | --- |
| GpuAdvancedSubtensor1 | Yes | ? | ? |  |
| GpuAdvancedIncSubtensor1 | Yes | Yes | No |  |
| GpuAdvancedIncSubtensor1_dev20 | Yes | no | Yes |  |
| GpuShape | Yes | Prbably | ? |  |
| SpecifyShape | Yes | Yes | ? |  |
| GpuContiguous | Yes | Yes | Yes |  |
| GpuEye | Yes | Yes | Yes |  |
| GpuSoftmax | Yes | yes | Yes |  |
| GpuGer | Yes | Yes | No |  |
| GpuImages2Neibs | Yes | Yes | No |  |
- [x] [GpuCrossentropySoftmaxArgmax1HotWithBias](gh-3163) got speed up. Need to port the speed up
- [x] cudnn ops
- [x] opt to move assert to new back-end.
- [x] GpuDot22Scalar
- [x] GpuCumSum (#4339)
- [x] GpuMultinomialFromUniform, float16: yes
- [x] BatchedDotOp
- [x] optimization - gpu_elemwise_alloc and tensor.opt.local_useless_elemwise needed by it. this was making cuda tests for opt elemwise_alloc fail. Make sure those tests are in gpuarray.
- [x] PrintOp
- [x] not needed - opt local_gpu_dot_to_dot22 (for optimizer=fast_compile) CPU Dot to gemm
- [x] lazy ifelse
- [x] max_inputs_to_GpuElemwise, split_huge_add_or_mul
- [x] GpuSplit (easy)
- [x] local_gpu_extract_diagonal
- [x] ErfinvGPU
- [x]  corrmm
- [x] GpuDownsampleFactorMaxGradGrad
- [x] gpu_solve
- [ ] device_properties, we currently use the one in the old back-end update to use the new one. We can change the interface, but the information should be available and documented all at one place.

Probably never unless they get needed:
- [ ] [Inc]DiagonalSubtensor (opt only, as the python code work for numpy.ndarray and the old CudaNdarray)
- [ ] CURAND_Base
- [ ]  contiguous_neg_strides
- [ ] CNDA_BEGIN_ALLOW_THREADS, CNDA_END_ALLOW_THREADS

Other Theano op that I think we shouldn't convert: GpuConv3D, GpuConvGrad3D, GpuConvTransp3D

**Comment by lamblin** [comment]
We should also add the CuDNN Ops to that list.

**Comment by nouiz** [comment]
It is already there (or you added it).

**Comment by Sentient07** [comment]
I think `Flatten` is missing here. When i use the existing cpu implementation of flatten on a shared variable on GPU, i get a `TensorType` Variable. The current workaround as @lamblin said is using GpuReshape. I don't think we'd have this problem if we port Flatten to GPU

**Comment by lamblin** [comment]
We deliberately decided not to use Flatten any more, because it is only a special case of Reshape.
These cases should be handled by having helper function combining Ops together, not having more Ops.

**Comment by Dapid** [comment]
What is the status of the convolutional operations on fp16? I am mostly interested in half precision storage, since I have fairly big volumetric data.

Can I help with it, or does it need deep knowledge of the internals?

**Comment by lamblin** [comment]
All that was needed is ported already.
